File: clause/condition.go
package clause

// A condition or predicate.
// Conditions can be used in a variety of SQL clauses. They're mainly used in a
// Select statement's WHERE clause.
type Condition interface {
	// Performs a logical AND operation between this Condition and the specified expression.
	//
	// The expression to be logically ANDed with this Condition.
	//
	// return a new Condition representing the logical AND operation.
	And(expr *Expression) *Condition

	// Performs a logical OR operation between this Condition and the specified expression.
	//
	// The expression to be logically ORed with this Condition.
	//
	// return a new Condition representing the logical OR operation.
	Or(expr *Expression) *Condition
}

// An operator used for combining conditions.
type Operator struct {
	keyword string
}

var (
	And Operator = Operator{"and"}
	Or  Operator = Operator{"or"}
)

// The logical expression.
//
// Combine this condition with another condition using the specified operator.
type Logical struct {
	Left     *Condition
	Right    *Condition
	Operator *Operator
}

// A comparator to be used in conditions to form comparison predicates.
type Comparator struct {
	keyword string
}

var (
	Eq         Comparator = Comparator{"eq"}
	Ne         Comparator = Comparator{"ne"}
	Lt         Comparator = Comparator{"lt"}
	Gt         Comparator = Comparator{"gt"}
	Le         Comparator = Comparator{"le"}
	Ge         Comparator = Comparator{"ge"}
	In         Comparator = Comparator{"in"}
	NotIn      Comparator = Comparator{"not in"}
	Like       Comparator = Comparator{"like"}
	NotLike    Comparator = Comparator{"not like"}
	Between    Comparator = Comparator{"between"}
	NotBetween Comparator = Comparator{"not between"}
	IsNull     Comparator = Comparator{"is null"}
	IsNotNull  Comparator = Comparator{"is not null"}
)

// The comparison expression.
type Comparison struct {
	Left     *Expression
	Operator *Comparator
	Right    *Expression
}

File: src/main/java/io/github/drawmoon/saber/Condition.java
/*
 *            _
 *  ___  __ _| |__   ___ _ __
 * / __|/ _` | '_ \ / _ \ '__|
 * \__ \ (_| | |_) |  __/ |
 * |___/\__,_|_.__/ \___|_|
 *
 * Copyright 2024 drash
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.github.drawmoon.saber;

/**
 * Represents a condition or predicate in SQL queries.
 *
 * <p>Conditions are used in various SQL clauses, primarily in the WHERE clause of a SELECT
 * statement. They can be combined using logical operators to form complex conditions.
 *
 * @author drash
 * @version 1.0
 * @since 2024
 */
public interface Condition extends Expression {

  /**
   * Performs a logical AND operation between this Condition and the specified expression.
   *
   * @param expr The expression to be logically ANDed with this Condition.
   * @return return a new Condition representing the logical AND operation.
   */
  Condition and(Expression expr);

  /**
   * Performs a logical OR operation between this Condition and the specified expression.
   *
   * @param expr The expression to be logically ORed with this Condition.
   * @return return a new Condition representing the logical OR operation.
   */
  Condition or(Expression expr);
}
